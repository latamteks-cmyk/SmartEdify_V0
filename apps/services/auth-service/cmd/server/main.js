import express from 'express';

import { registerHandler } from './internal/adapters/http/register.handler';
// ...import otros handlers

const app = express();
app.use(express.json());

// Endpoints principales
app.post('/register', registerHandler);
import { loginHandler } from '../../internal/adapters/http/login.handler';
app.post('/login', loginHandler);
import { logoutHandler } from '../../internal/adapters/http/logout.handler';
app.post('/logout', logoutHandler);
import { forgotPasswordHandler } from '../../internal/adapters/http/forgot-password.handler';
app.post('/forgot-password', forgotPasswordHandler);
import { resetPasswordHandler } from '../../internal/adapters/http/reset-password.handler';
app.post('/reset-password', resetPasswordHandler);
import { rolesHandler, permissionsHandler } from '../../internal/adapters/http/roles-permissions.handler';
app.get('/roles', rolesHandler);
app.get('/permissions', permissionsHandler);

const PORT = process.env.AUTH_PORT || 8080;
app.listen(PORT, () => {
  console.log(`Auth Service escuchando en puerto ${PORT}`);
});
