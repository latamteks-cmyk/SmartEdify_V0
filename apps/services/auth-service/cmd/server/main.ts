/*
 * Copyright (c) 2024 SmartEdify contributors
 * Licensed under the MIT License. See the LICENSE file in the project root for details.
 */

import 'dotenv/config';
import { parseEnv } from '../../internal/config/env';
import { startTracing, shutdownTracing } from '../../internal/observability/tracing';

import { context, trace } from '@opentelemetry/api';

// Validación de entorno temprana (fail-fast en producción)
const __env = parseEnv(process.env);
if (__env.NODE_ENV === 'production') {
  if (!__env.AUTH_ADMIN_API_KEY) {
    // Asegurar que los endpoints administrativos estén protegidos en prod
    throw new Error('AUTH_ADMIN_API_KEY es obligatorio en producción');
  }
}

// Inicializar tracing (no bloquear si falla)
startTracing().catch(err => {
  console.error('[tracing] init failed', err);
});

import { randomUUID } from 'crypto';

import express from 'express';
import pino from 'pino';
import pinoHttp from 'pino-http';
import client from 'prom-client';

import * as pgAdapter from '@db/pg.adapter';
import { forgotPasswordHandler } from '../../internal/adapters/http/forgot-password.handler';
import { loginHandler } from '../../internal/adapters/http/login.handler';
import { logoutHandler } from '../../internal/adapters/http/logout.handler';
import { refreshHandler } from '../../internal/adapters/http/refresh.handler';
import { registerHandler } from '../../internal/adapters/http/register.handler';
import { resetPasswordHandler } from '../../internal/adapters/http/reset-password.handler';
import { rolesHandler, permissionsHandler } from '../../internal/adapters/http/roles-permissions.handler';
import { authorizeHandler } from '../../internal/adapters/http/authorize.handler';
import { tokenHandler } from '../../internal/adapters/http/token.handler';
import { userinfoHandler } from '../../internal/adapters/http/userinfo.handler';
import { introspectionHandler } from '../../internal/adapters/http/introspection.handler';
import { revocationHandler } from '../../internal/adapters/http/revocation.handler';
import { openIdConfigurationHandler } from '../../internal/adapters/http/openid-configuration.handler';
import { redisPing } from '../../internal/adapters/redis/redis.adapter';
import { loginRateLimiter, bruteForceGuard, adminRateLimiter } from '../../internal/middleware/rate-limit';
import { adminAuthMiddleware } from '../../internal/middleware/admin-auth';
import { getPublicJwks, rotateKeys, getCurrentKey } from '../../internal/security/keys';
import { revokeSessionsByKid } from '../../internal/security/jwt';

const rootLogger = pino({
  level: process.env.AUTH_LOG_LEVEL || 'info',
  base: { service: 'auth-service' },
  timestamp: pino.stdTimeFunctions.isoTime
});
// En test a veces pino-http se rompe por symbol collisions; proveemos fallback plano
const logger: any = process.env.NODE_ENV === 'test'
  ? {
      info: (...a: any[]) => { if (process.env.AUTH_TEST_LOGS) console.log('[info]', ...a); },
      error: (...a: any[]) => { if (process.env.AUTH_TEST_LOGS) console.error('[error]', ...a); },
      warn: (...a: any[]) => { if (process.env.AUTH_TEST_LOGS) console.warn('[warn]', ...a); },
      debug: (..._a: any[]) => {},
      child: () => this
    }
  : rootLogger.child({});

// Diagnóstico configuración PG
logger.debug({
  pg: {
    host: process.env.PGHOST,
    port: process.env.PGPORT,
    user: process.env.PGUSER,
    db: process.env.PGDATABASE
  }
}, 'PG environment variables');

// Prometheus metrics setup
const register = new client.Registry();
// Nota: collectDefaultMetrics devuelve void en prom-client v15; simplemente no lo llamamos en test
if (process.env.NODE_ENV !== 'test') {
  client.collectDefaultMetrics({ register });
}

// Custom metrics
const httpRequestsTotal = new client.Counter({
  name: 'auth_http_requests_total',
  help: 'Total de requests HTTP recibidas',
  labelNames: ['method', 'route', 'status']
});
const httpRequestDuration = new client.Histogram({
  name: 'auth_http_request_duration_seconds',
  help: 'Duración de requests HTTP en segundos',
  labelNames: ['method', 'route', 'status'],
  buckets: [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2, 5]
});
register.registerMetric(httpRequestsTotal);
register.registerMetric(httpRequestDuration);

// Business metrics (login, password reset)
export const loginSuccessCounter = new client.Counter({
  name: 'auth_login_success_total',
  help: 'Total de logins exitosos'
});
export const loginFailCounter = new client.Counter({
  name: 'auth_login_fail_total',
  help: 'Total de logins fallidos (credenciales inválidas)'
});
export const passwordResetRequestedCounter = new client.Counter({
  name: 'auth_password_reset_requested_total',
  help: 'Total de solicitudes de recuperación de contraseña generadas'
});
export const passwordResetCompletedCounter = new client.Counter({
  name: 'auth_password_reset_completed_total',
  help: 'Total de recuperaciones de contraseña completadas'
});
register.registerMetric(loginSuccessCounter);
register.registerMetric(loginFailCounter);
register.registerMetric(passwordResetRequestedCounter);
register.registerMetric(passwordResetCompletedCounter);

// Métricas de rotación (refresh)
export const refreshRotatedCounter = new client.Counter({
  name: 'auth_refresh_rotated_total',
  help: 'Total de rotaciones exitosas de refresh token'
});
export const refreshReuseBlockedCounter = new client.Counter({
  name: 'auth_refresh_reuse_blocked_total',
  help: 'Intentos bloqueados de reutilización de refresh token rotado'
});
register.registerMetric(refreshRotatedCounter);
register.registerMetric(refreshReuseBlockedCounter);

export const tokenRevokedCounter = new client.Counter({
  name: 'auth_token_revoked_total',
  help: 'Total de tokens revocados por logout o acciones administrativas',
  labelNames: ['type']
});
register.registerMetric(tokenRevokedCounter);

// JWKS metrics
export const jwksKeysTotal = new client.Gauge({
  name: 'auth_jwks_keys_total',
  help: 'Número de claves activas (current + next + retiring)',
  labelNames: ['status']
});
export const jwksRotationCounter = new client.Counter({
  name: 'auth_jwks_rotation_total',
  help: 'Total de rotaciones manuales ejecutadas'
});
register.registerMetric(jwksKeysTotal);
register.registerMetric(jwksRotationCounter);

export const app = express();
// Export util para tests que permita limpiar intervalo si en algún momento se inicializa
export async function _cleanupMetrics() { /* noop actual (mantener API por si cambiamos) */ }
app.use(express.json());
app.use(express.urlencoded({ extended: false }));

// Middleware request-id
app.use((req, res, next) => {
  const rid = (req.headers['x-request-id'] as string) || randomUUID();
  (req as any).requestId = rid;
  res.setHeader('x-request-id', rid);
  next();
});

// pino-http para logs estructurados
// Uso básico de pino-http (dejamos que cree su propia instancia enlazada)
if (process.env.NODE_ENV !== 'test') {
  app.use(pinoHttp({
    customProps: (req: any) => {
      const span = trace.getSpan(context.active());
      const sc = span?.spanContext();
      return {
        requestId: (req as any).requestId,
        trace_id: sc?.traceId,
        span_id: sc?.spanId
      };
    },
    logger: logger as any
  }));
}

// Metrics middleware (después de logging para capturar status final)
app.use((req, res, next) => {
  const startHr = process.hrtime();
  res.on('finish', () => {
    const diff = process.hrtime(startHr);
    const duration = diff[0] + diff[1] / 1e9; // seconds
    const route = (req.route && req.route.path) || req.path || 'unmatched';
    const labels = { method: req.method, route, status: String(res.statusCode) } as any;
    httpRequestsTotal.inc(labels);
    httpRequestDuration.observe(labels, duration);
  });
  next();
});

// Health endpoint
app.get('/health', async (req, res) => {
  const start = Date.now();
  let dbOk = false;
  let redisOk = false;
  try {
  await pgAdapter.pool.query('SELECT 1');
    dbOk = true;
  } catch (e) {
    logger.error({ err: e }, 'DB health check failed');
  }
  try {
    const pong = await redisPing();
    redisOk = pong === 'PONG';
  } catch (e) {
    logger.error({ err: e }, 'Redis health check failed');
  }
  const status = dbOk && redisOk ? 'ok' : 'degraded';
  const body = { status, db: dbOk, redis: redisOk, uptime_s: process.uptime(), latency_ms: Date.now() - start };
  // En entorno de test devolvemos 200 siempre para estabilidad de contratos aislados
  const code = process.env.NODE_ENV === 'test' ? 200 : (status === 'ok' ? 200 : 503);
  res.status(code).json(body);
});

// Metrics endpoint
app.get('/metrics', async (req, res) => {
  try {
    res.setHeader('Content-Type', register.contentType);
    res.end(await register.metrics());
  } catch (e) {
    logger.error({ err: e }, 'Error generando métricas');
    res.status(500).end();
  }
});

// JWKS público
app.get('/.well-known/jwks.json', async (_req, res) => {
  try {
    const jwks = await getPublicJwks();
    // Actualizar gauge por estado
    const counts: Record<string, number> = {};
    for (const k of jwks.keys) {
      counts[k.status] = (counts[k.status] || 0) + 1;
    }
    const statuses = ['current','next','retiring'];
    for (const st of statuses) {
      jwksKeysTotal.set({ status: st }, counts[st] || 0);
    }
    res.json(jwks);
  } catch (e: any) {
    logger.error({ err: e }, 'Error obteniendo JWKS');
    res.status(500).json({ error: 'jwks_error' });
  }
});

app.get('/.well-known/openid-configuration', openIdConfigurationHandler);

// Rotación manual (MVP) - proteger en producción
app.post('/admin/rotate-keys', adminRateLimiter, adminAuthMiddleware, async (_req, res) => {
  try {
    const result = await rotateKeys();
    jwksRotationCounter.inc();
    res.json({ message: 'rotated', current: { kid: result.newCurrent.kid }, next: result.newNext ? { kid: result.newNext.kid } : null });
  } catch (e: any) {
    logger.error({ err: e }, 'Error en rotación manual');
    res.status(500).json({ error: 'rotation_failed' });
  }
});

app.post('/admin/revoke-kid', adminRateLimiter, adminAuthMiddleware, async (req, res) => {
  const kid = typeof req.body?.kid === 'string' ? req.body.kid.trim() : '';
  if (!kid) {
    return res.status(400).json({ error: 'kid_required' });
  }
  try {
    const result = await revokeSessionsByKid(kid);
    res.json({ message: 'revoked', ...result });
  } catch (e: any) {
    logger.error({ err: e, kid }, 'Error revocando sesiones por kid');
    res.status(500).json({ error: 'revoke_failed' });
  }
});

if (process.env.NODE_ENV !== 'production') {
  app.get('/debug/current-kid', async (_req, res) => {
    const key = await getCurrentKey();
    res.json({ kid: key.kid });
  });
}

// Endpoints principales
app.get('/authorize', authorizeHandler);
app.post('/token', tokenHandler);
app.get('/userinfo', userinfoHandler);
app.post('/introspection', introspectionHandler);
app.post('/revocation', revocationHandler);
app.post('/register', registerHandler);
app.post('/login', loginRateLimiter, bruteForceGuard, loginHandler);
app.post('/logout', logoutHandler);
app.post('/forgot-password', forgotPasswordHandler);
app.post('/reset-password', resetPasswordHandler);
app.post('/refresh-token', refreshHandler);
app.get('/roles', rolesHandler);
app.get('/permissions', permissionsHandler);

if (process.env.NODE_ENV !== 'test') {
  const PORT = process.env.AUTH_PORT || 8080;
  const server = app.listen(PORT, () => {
    logger.info({ msg: 'Auth Service escuchando', port: PORT });
  });
  const graceful = async () => {
    try { await shutdownTracing(); } catch {}
    server.close(() => process.exit(0));
  };
  process.on('SIGINT', graceful);
  process.on('SIGTERM', graceful);
}
