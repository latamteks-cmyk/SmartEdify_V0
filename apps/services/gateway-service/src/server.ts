import express from 'express';
import helmet from 'helmet';
import { config } from './config/env';
import { corsMiddleware } from './middleware/cors';
import { loggingMiddleware, requestIdMiddleware } from './middleware/logging';
import { tracingMiddleware } from './middleware/tracing';
import { initializeTracing } from './config/tracing';
import routes from './routes/index';
import { services } from './config/services';
import { httpMetricsMiddleware, metricsRouter } from './observability/metrics';

const app = express();

// Initialize tracing
initializeTracing().catch(err => {
  console.error('Failed to initialize tracing:', err);
});

// Security middleware
const helmetCspDirectives: any = {
  defaultSrc: ["'self'"],
  styleSrc: ["'self'", ...(config.NODE_ENV === 'production' ? [] : ["'unsafe-inline'"])],
  scriptSrc: ["'self'"],
  imgSrc: ["'self'", 'data:', 'https:'],
};

app.use(
  helmet({
    contentSecurityPolicy: { directives: helmetCspDirectives },
    hsts: { maxAge: 31536000, includeSubDomains: true, preload: true },
    frameguard: { action: 'deny' },
    referrerPolicy: { policy: 'no-referrer' },
  })
);

// Request processing middleware
app.use(requestIdMiddleware);
app.use(tracingMiddleware);
app.use(httpMetricsMiddleware);
app.use(loggingMiddleware);
app.use(corsMiddleware);
// Ensure CORS headers are present even without Origin header
app.use((req, res, next) => {
  if (!res.getHeader('Access-Control-Allow-Origin')) {
    res.setHeader('Access-Control-Allow-Origin', config.CORS_ORIGIN.includes('*') ? '*' : config.CORS_ORIGIN.split(',')[0]);
  }
  if (!res.getHeader('Access-Control-Allow-Credentials')) {
    res.setHeader('Access-Control-Allow-Credentials', String(config.CORS_CREDENTIALS));
  }
  next();
});
app.use(express.json({ limit: '10mb' }));
app.use(express.urlencoded({ extended: true, limit: '10mb' }));

// Rate limiting will be applied per route as needed

// Trust proxy for accurate IP addresses
app.set('trust proxy', 1);

// Metrics route (before other routes to avoid interference)
app.use(metricsRouter);

// Routes
app.use('/', routes);

// Global error handler
app.use((err: Error, req: express.Request, res: express.Response, next: express.NextFunction) => {
  console.error('Unhandled error:', err);
  
  if (res.headersSent) {
    return next(err);
  }
  
  res.status(500).json({
    error: 'Internal server error',
    requestId: req.headers['x-request-id'],
    code: 'INTERNAL_ERROR'
  });
});

// Only start server if not in test environment
if (config.NODE_ENV !== 'test') {
  const server = app.listen(config.PORT, () => {
    console.log(`ðŸš€ Gateway Service running on port ${config.PORT}`);
    console.log(`ðŸ“Š Environment: ${config.NODE_ENV}`);
    console.log(`ðŸ”— Services configured: ${Object.keys(services).join(', ')}`);
  });

  // Handle graceful shutdown
  process.on('SIGTERM', () => {
    console.log('SIGTERM received, shutting down gracefully');
    server.close(() => {
      console.log('Process terminated');
      process.exit(0);
    });
  });

  process.on('SIGINT', () => {
    console.log('SIGINT received, shutting down gracefully');
    server.close(() => {
      console.log('Process terminated');
      process.exit(0);
    });
  });
}

export default app;
