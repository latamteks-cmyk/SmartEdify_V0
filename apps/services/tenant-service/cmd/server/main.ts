/// <reference types="../..//types/fastify-di" />
// Copyright (c) 2024 SmartEdify contributors
// Licensed under the MIT License. See the LICENSE file in the project root for details.

import Fastify from 'fastify';
import { config } from '../../internal/config/env.js';
import { pool } from '../../internal/adapters/repo/db.js';
import { PgTenantRepository } from '../../internal/adapters/repo/tenant-pg.js';
import { PgUnitRepository } from '../../internal/adapters/repo/unit-pg.js';
import { PgMembershipRepository } from '../../internal/adapters/repo/membership-pg.js';
import { PgGovernanceRepository } from '../../internal/adapters/repo/governance-pg.js';
import { PgRolesRepository } from '../../internal/adapters/repo/roles-pg.js';
import { healthRoute } from '../../internal/adapters/http/routes/health.js';
import { metricsRoute } from '../../internal/adapters/http/routes/metrics.js';
import { tenantRoutes } from '../../internal/adapters/http/routes/tenants.js';
import { unitRoutes } from '../../internal/adapters/http/routes/units.js';
import { membershipRoutes } from '../../internal/adapters/http/routes/memberships.js';
import { governanceRoutes } from '../../internal/adapters/http/routes/governance.js';
import { tenantContextRoute } from '../../internal/adapters/http/routes/tenant-context.js';
import { outboxDlqRoutes } from '../../internal/adapters/http/routes/outbox-dlq.js';
import { PgOutboxRepository } from '../../internal/adapters/repo/outbox-pg.js';
import { OutboxPoller } from '../../internal/adapters/publisher/outbox-poller.js';
import { LoggingPublisher } from '../../internal/adapters/publisher/publisher.js';
import { KafkaPublisher } from '../../internal/adapters/publisher/kafka-publisher.js';
import { LoggingConsumer } from '../../internal/adapters/consumer/consumer.js';
import { KafkaLagConsumer } from '../../internal/adapters/consumer/kafka-consumer.js';
import { RabbitMqPublisher } from '../../internal/adapters/publisher/rabbitmq-publisher.js';
import { RabbitMqDeadLetterConsumer } from '../../internal/adapters/consumer/rabbitmq-dead-letter-consumer.js';
import { authJwtPlugin } from '../../internal/adapters/http/plugins/auth-jwt.js';
import { startTracing, shutdownTracing } from '../../internal/observability/tracing.js';
import { JwksProvider } from '../../internal/adapters/security/jwks-provider.js';

async function build() {
  await startTracing();
  const app = Fastify({
    logger: {
      level: config.logLevel,
      transport: process.env.NODE_ENV === 'production' ? undefined : {
        target: 'pino-pretty',
        options: { translateTime: 'HH:MM:ss', ignore: 'pid,hostname' }
      }
    }
  });

  // Simple DI container
  const di = {
    tenantRepo: new PgTenantRepository(),
    unitRepo: new PgUnitRepository(),
    membershipRepo: new PgMembershipRepository(),
    governanceRepo: new PgGovernanceRepository(),
    rolesRepo: new PgRolesRepository(),
    outboxRepo: new PgOutboxRepository()
  } as const;
  app.decorate('di', di);

  let publisher: any;
  if (config.publisherKind === 'kafka' && config.kafkaBrokers.length > 0) {
    publisher = new KafkaPublisher({
      brokers: config.kafkaBrokers,
      clientId: config.kafkaClientId,
      topicPrefix: config.kafkaTopicPrefix,
      acks: config.kafkaAcks,
      logger: app.log
    });
    app.log.info({ brokers: config.kafkaBrokers }, 'KafkaPublisher enabled');
  } else if (config.publisherKind === 'rabbitmq') {
    publisher = new RabbitMqPublisher({
      url: config.rabbitUrl,
      exchange: config.rabbitExchange,
      routingKey: config.rabbitRoutingKey,
      queue: config.rabbitQueue,
      deadLetterExchange: config.rabbitDeadLetterExchange,
      deadLetterQueue: config.rabbitDeadLetterQueue,
      deadLetterRoutingKey: config.rabbitDeadLetterRoutingKey,
      logger: app.log
    });
    app.log.info({ exchange: config.rabbitExchange, queue: config.rabbitQueue }, 'RabbitMqPublisher enabled');
  } else {
    publisher = new LoggingPublisher(app.log);
    if (config.publisherKind === 'kafka') {
      app.log.warn('Kafka requested but no brokers configured, falling back to LoggingPublisher');
    } else if (config.publisherKind === 'rabbitmq') {
      app.log.warn('RabbitMQ requested but no configuration provided, falling back to LoggingPublisher');
    }
  }
  const poller = new OutboxPoller(di.outboxRepo, publisher, { intervalMs: config.outboxPollIntervalMs, batchSize: config.outboxBatchSize, logger: app.log });
  ;(app as any).publisher = publisher; // exposición para health route
  poller.start();

  // Consumer (solo lag por ahora)
  let consumer: any = null;
  if (config.consumerKind === 'kafka' && config.kafkaBrokers.length > 0) {
    consumer = new KafkaLagConsumer({
      brokers: config.kafkaBrokers,
      clientId: config.kafkaClientId + '-lag',
      groupId: config.kafkaConsumerGroupId,
      topicPrefix: config.kafkaTopicPrefix,
      lagIntervalMs: config.kafkaConsumerLagIntervalMs,
      logger: app.log
    });
    consumer.start().catch((e: any) => app.log.error(e, 'consumer start failed'));
  } else if (config.consumerKind === 'rabbitmq') {
    consumer = new RabbitMqDeadLetterConsumer({
      url: config.rabbitUrl,
      exchange: config.rabbitExchange,
      routingKey: config.rabbitRoutingKey,
      queue: config.rabbitQueue,
      deadLetterExchange: config.rabbitDeadLetterExchange,
      deadLetterQueue: config.rabbitDeadLetterQueue,
      deadLetterRoutingKey: config.rabbitDeadLetterRoutingKey,
      checkIntervalMs: config.rabbitDeadLetterCheckIntervalMs,
      alertThreshold: config.rabbitDeadLetterAlertThreshold,
      logger: app.log
    });
    consumer.start().catch((e: any) => app.log.error(e, 'rabbitmq dead-letter consumer failed'));
  } else if (config.consumerKind === 'logging') {
    consumer = new LoggingConsumer();
    await consumer.start();
  }

  await app.register(metricsRoute);
  await app.register(healthRoute);

  // Registrar plugin de auth: se prioriza JWKS si está configurado; fallback a clave estática
  let jwksProvider: JwksProvider | undefined;
  if (config.jwksUrl) {
    jwksProvider = new JwksProvider({ url: config.jwksUrl, cacheTtlMs: config.jwksCacheTtlMs, logger: app.log });
  }
  const publicKey = process.env.TENANT_JWT_PUBLIC_KEY?.replace(/\\n/g, '\n');
  await app.register(authJwtPlugin, { publicKeyPem: publicKey, jwksProvider, required: false });

  await app.register(tenantRoutes);
  await app.register(unitRoutes);
  await app.register(membershipRoutes);
  await app.register(governanceRoutes);
  await app.register(tenantContextRoute);
  await app.register(outboxDlqRoutes);

  app.addHook('onClose', async () => {
    await pool.end();
    await poller.stop();
    if (consumer) { await consumer.shutdown().catch(() => {}); }
    await shutdownTracing().catch(e => app.log.error(e, 'tracing shutdown error'));
  });

  return app;
}

build().then(app => {
  app.listen({ port: config.port, host: '0.0.0.0' })
    .then(() => app.log.info(`tenant-service listening on ${config.port}`))
    .catch(err => {
      app.log.error(err, 'startup error');
      process.exit(1);
    });
});
