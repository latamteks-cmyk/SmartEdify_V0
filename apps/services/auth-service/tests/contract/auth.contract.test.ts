import request from 'supertest';
import { app } from '../../cmd/server/main';
import { contractSnapshot } from './utils/snapshot';

const dbMock = require('../../internal/adapters/db/pg.adapter') as any;
const redisModule = require('../../internal/adapters/redis/redis.adapter');
const redis = redisModule.default;

const TENANT_ID = 'default';
const PASSWORD = 'ContractPass!123';

async function resetState() {
  if (typeof dbMock.__resetMock === 'function') {
    dbMock.__resetMock();
  }
  if (redis && typeof redis.flushdb === 'function') {
    await redis.flushdb();
  }
  const stores = ['__REFRESH_STORE__', '__AUTH_CODE_STORE__', '__PWDRESET_STORE__'];
  for (const key of stores) {
    const store = (global as any)[key];
    if (store && typeof store.clear === 'function') {
      store.clear();
    }
  }
}

async function registerUser(email: string, name = 'Contract User') {
  return request(app)
    .post('/register')
    .send({ email, password: PASSWORD, name, tenant_id: TENANT_ID });
}

describe('Auth contract snapshots', () => {
  beforeEach(async () => {
    process.env.AUTH_ADMIN_API_KEY = process.env.AUTH_ADMIN_API_KEY || 'test-admin-key';
    await resetState();
  });

  it('POST /register devuelve contrato estable', async () => {
    const response = await registerUser('contract.register@demo.com', 'Register User');
    expect(response.status).toBe(201);
    expect(contractSnapshot(response)).toMatchSnapshot('POST /register 201');
  });

  it('POST /login devuelve tokens en formato esperado', async () => {
    const email = 'contract.login@demo.com';
    await registerUser(email, 'Login User');
    const response = await request(app)
      .post('/login')
      .send({ email, password: PASSWORD, tenant_id: TENANT_ID });
    expect(response.status).toBe(200);
    expect(contractSnapshot(response)).toMatchSnapshot('POST /login 200');
  });

  it('POST /refresh-token rota el refresh token sin romper contrato', async () => {
    const email = 'contract.refresh@demo.com';
    await registerUser(email, 'Refresh User');
    const loginRes = await request(app)
      .post('/login')
      .send({ email, password: PASSWORD, tenant_id: TENANT_ID });
    expect(loginRes.status).toBe(200);
    const refreshToken: string | undefined = loginRes.body?.refresh_token;
    expect(typeof refreshToken).toBe('string');
    const response = await request(app)
      .post('/refresh-token')
      .send({ refresh_token: refreshToken });
    expect(response.status).toBe(200);
    expect(contractSnapshot(response)).toMatchSnapshot('POST /refresh-token 200');
  });

  it('GET /health reporta dependencias principales', async () => {
    const response = await request(app).get('/health');
    expect(response.status).toBe(200);
    expect(contractSnapshot(response)).toMatchSnapshot('GET /health 200');
  });
});
