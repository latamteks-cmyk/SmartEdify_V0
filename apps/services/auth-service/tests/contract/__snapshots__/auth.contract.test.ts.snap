exports[`Auth contract snapshots POST /register devuelve contrato estable POST /register 201 1`] = `
{
  "status": 201,
  "headers": {
    "content-type": "application/json; charset=utf-8",
    "x-request-id": "<x-request-id>"
  },
  "body": {
    "message": "Usuario registrado",
    "user": {
      "id": "<id>",
      "email": "contract.register@demo.com",
      "name": "Register User",
      "roles": [
        "user"
      ]
    }
  }
}
`;

exports[`Auth contract snapshots POST /login devuelve tokens en formato esperado POST /login 200 1`] = `
{
  "status": 200,
  "headers": {
    "content-type": "application/json; charset=utf-8",
    "x-request-id": "<x-request-id>"
  },
  "body": {
    "message": "Login exitoso",
    "access_token": "<access_token>",
    "refresh_token": "<refresh_token>",
    "token_type": "Bearer",
    "expires_in": "<expires_in>",
    "roles": [
      "user"
    ]
  }
}
`;

exports[`Auth contract snapshots POST /refresh-token rota el refresh token sin romper contrato POST /refresh-token 200 1`] = `
{
  "status": 200,
  "headers": {
    "content-type": "application/json; charset=utf-8",
    "x-request-id": "<x-request-id>"
  },
  "body": {
    "access_token": "<access_token>",
    "refresh_token": "<refresh_token>",
    "token_type": "Bearer",
    "expires_in": "<expires_in>"
  }
}
`;

exports[`Auth contract snapshots GET /health reporta dependencias principales GET /health 200 1`] = `
{
  "status": 200,
  "headers": {
    "content-type": "application/json; charset=utf-8",
    "x-request-id": "<x-request-id>"
  },
  "body": {
    "status": "ok",
    "db": true,
    "redis": true,
    "uptime_s": "<uptime_s>",
    "latency_ms": "<latency_ms>"
  }
}
`;
